File: server/lib/preload.ts
/**
 * Background preload — refresh all filter combinations every N minutes so
 * console filter changes are served instantly from memory/disk cache.
 * AI summaries are NOT preloaded; those always go through Supervisor on demand.
 */
import {
  getFreshMetricsBundle,
  getMemoryCacheStats,
  refreshMetricsBundle,
  runAnalyticsQuery,
  warmMemoryCacheFromDisk,
} from './analytics.js';
import { buildFilterOptions } from './metricsTransform.js';
import { sqlConfigured } from './databricksSql.js';

const INTERVAL_MS = Number(process.env.PRELOAD_INTERVAL_MINUTES || 15) * 60 * 1000;
const CONCURRENCY = Math.max(1, Number(process.env.PRELOAD_CONCURRENCY || 3));
const DEFAULT_YEAR = String(process.env.PRELOAD_DEFAULT_YEAR || '2026');

export function preloadEnabled(): boolean {
  return String(process.env.PRELOAD_ENABLED ?? 'true').toLowerCase() !== 'false';
}

export interface PreloadStatus {
  enabled: boolean;
  running: boolean;
  interval_minutes: number;
  concurrency: number;
  lastRunStartedAt: number | null;
  lastRunFinishedAt: number | null;
  lastRunError: string | null;
  combinationsTotal: number;
  combinationsDone: number;
  combinationsSkipped: number;
  memoryCacheEntries: number;
  nextRunAt: number | null;
}

const status: PreloadStatus = {
  enabled: preloadEnabled(),
  running: false,
  interval_minutes: INTERVAL_MS / 60000,
  concurrency: CONCURRENCY,
  lastRunStartedAt: null,
  lastRunFinishedAt: null,
  lastRunError: null,
  combinationsTotal: 0,
  combinationsDone: 0,
  combinationsSkipped: 0,
  memoryCacheEntries: 0,
  nextRunAt: null,
};

let schedulerStarted = false;
let nextRunTimer: ReturnType<typeof setTimeout> | null = null;

async function loadFilterCombinations(): Promise<Record<string, unknown>[]> {
  const { rows } = await runAnalyticsQuery('dashboard_filter_options', {});
  const options = buildFilterOptions(rows as Record<string, unknown>[]);
  const years = options.years.length ? options.years.map(String) : [DEFAULT_YEAR];
  const sites: (string | null)[] = [null, ...options.sites];
  const regionSets: (string[] | null)[] = [null, ...options.regions.map(r => [r])];

  const combos: Record<string, unknown>[] = [];
  for (const year of years) {
    for (const site of sites) {
      for (const region of regionSets) {
        combos.push({
          year,
          site: site || 'All',
          region: region || [],
          timeframe: 'Week',
        });
      }
    }
  }

  combos.sort((a, b) => {
    const score = (c: Record<string, unknown>) => {
      let s = 0;
      if (String(c.year) === DEFAULT_YEAR) s += 100;
      if (c.site === 'All') s += 50;
      if (!c.region || (Array.isArray(c.region) && c.region.length === 0)) s += 25;
      return s;
    };
    return score(b) - score(a);
  });

  return combos;
}

async function runPool<T>(
  items: T[],
  concurrency: number,
  worker: (item: T, index: number) => Promise<void>,
): Promise<void> {
  let index = 0;
  async function next(): Promise<void> {
    while (index < items.length) {
      const i = index++;
      await worker(items[i], i);
    }
  }
  await Promise.all(Array.from({ length: Math.min(concurrency, items.length) }, () => next()));
}

export async function runPreloadCycle(): Promise<void> {
  if (!preloadEnabled() || !sqlConfigured()) return;
  if (status.running) {
    console.log('[preload] cycle already running — skip');
    return;
  }

  status.running = true;
  status.lastRunStartedAt = Date.now();
  status.lastRunError = null;
  status.combinationsDone = 0;
  status.combinationsSkipped = 0;

  try {
    const combos = await loadFilterCombinations();
    status.combinationsTotal = combos.length;
    console.log(`[preload] starting cycle — ${combos.length} combinations, concurrency=${CONCURRENCY}`);

    await runPool(combos, CONCURRENCY, async combo => {
      if (getFreshMetricsBundle(combo)) {
        status.combinationsSkipped++;
        status.combinationsDone++;
        return;
      }
      try {
        await refreshMetricsBundle(combo, msg => {
          if (status.combinationsDone % 10 === 0) {
            console.log(`[preload] ${status.combinationsDone}/${status.combinationsTotal} — ${msg}`);
          }
        });
      } catch (e) {
        console.warn('[preload] combo failed:', JSON.stringify(combo).slice(0, 80), (e as Error).message);
      }
      status.combinationsDone++;
    });

    status.memoryCacheEntries = getMemoryCacheStats().entries;
    status.lastRunFinishedAt = Date.now();
    const elapsed = ((status.lastRunFinishedAt - status.lastRunStartedAt!) / 1000).toFixed(0);
    console.log(
      `[preload] cycle complete — ${status.combinationsDone}/${status.combinationsTotal} ` +
      `(${status.combinationsSkipped} skipped fresh) in ${elapsed}s, ` +
      `${status.memoryCacheEntries} memory keys`,
    );
  } catch (e) {
    status.lastRunError = (e as Error).message;
    status.lastRunFinishedAt = Date.now();
    console.error('[preload] cycle failed:', status.lastRunError);
  } finally {
    status.running = false;
    status.memoryCacheEntries = getMemoryCacheStats().entries;
    scheduleNextRun();
  }
}

function scheduleNextRun(): void {
  if (!schedulerStarted || !preloadEnabled() || !sqlConfigured()) return;
  if (nextRunTimer) clearTimeout(nextRunTimer);
  status.nextRunAt = Date.now() + INTERVAL_MS;
  nextRunTimer = setTimeout(() => {
    void runPreloadCycle();
  }, INTERVAL_MS);
}

export function startPreloadScheduler(): void {
  if (schedulerStarted || !preloadEnabled() || !sqlConfigured()) return;
  schedulerStarted = true;
  status.enabled = true;
  warmMemoryCacheFromDisk();
  console.log(`[preload] scheduler started — every ${INTERVAL_MS / 60000} min, concurrency=${CONCURRENCY}`);
  void runPreloadCycle();
}

export function getPreloadStatus(): PreloadStatus {
  return {
    ...status,
    enabled: preloadEnabled(),
    interval_minutes: INTERVAL_MS / 60000,
    concurrency: CONCURRENCY,
    memoryCacheEntries: getMemoryCacheStats().entries,
  };
}

File: py_server/lib/preload.py
"""
Background preload — refresh all filter combinations every N minutes so
console filter changes are served instantly from memory/disk cache.
AI summaries are NOT preloaded; those always go through Supervisor on demand.
"""
from __future__ import annotations

import json
import os
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any, Callable, TypedDict

from py_server.lib.analytics import (
    get_fresh_metrics_bundle,
    get_memory_cache_stats,
    refresh_metrics_bundle,
    run_analytics_query,
    warm_memory_cache_from_disk,
)
from py_server.lib.databricks_sql import sql_configured
from py_server.lib.metrics_transform import build_filter_options

INTERVAL_MS = int(os.getenv('PRELOAD_INTERVAL_MINUTES') or 15) * 60 * 1000
CONCURRENCY = max(1, int(os.getenv('PRELOAD_CONCURRENCY') or 1))
DEFAULT_YEAR = os.getenv('PRELOAD_DEFAULT_YEAR') or '2026'
FULL_CYCLE_DELAY_MS = int(os.getenv('PRELOAD_FULL_DELAY_MINUTES') or 10) * 60 * 1000


def preload_enabled() -> bool:
    return (os.getenv('PRELOAD_ENABLED') or 'true').lower() != 'false'


class PreloadStatus(TypedDict):
    enabled: bool
    running: bool
    interval_minutes: float
    concurrency: int
    lastRunStartedAt: float | None
    lastRunFinishedAt: float | None
    lastRunError: str | None
    combinationsTotal: int
    combinationsDone: int
    combinationsSkipped: int
    memoryCacheEntries: int
    nextRunAt: float | None


_status: PreloadStatus = {
    'enabled': preload_enabled(),
    'running': False,
    'interval_minutes': INTERVAL_MS / 60000,
    'concurrency': CONCURRENCY,
    'lastRunStartedAt': None,
    'lastRunFinishedAt': None,
    'lastRunError': None,
    'combinationsTotal': 0,
    'combinationsDone': 0,
    'combinationsSkipped': 0,
    'memoryCacheEntries': 0,
    'nextRunAt': None,
}

_scheduler_started = False
_next_run_timer: threading.Timer | None = None
_status_lock = threading.Lock()


def _load_filter_combinations() -> list[dict[str, Any]]:
    result = run_analytics_query('dashboard_filter_options', {})
    rows = result.get('rows') or []
    options = build_filter_options(rows)
    years = [str(y) for y in options.get('years') or []] or [DEFAULT_YEAR]
    sites: list[str | None] = [None, *(options.get('sites') or [])]
    region_sets: list[list[str] | None] = [None]
    for r in options.get('regions') or []:
        region_sets.append([r])

    combos: list[dict[str, Any]] = []
    for year in years:
        for site in sites:
            for region in region_sets:
                combos.append({
                    'year': year,
                    'site': site or 'All',
                    'region': region or [],
                    'timeframe': 'FY',
                })

    def score(c: dict[str, Any]) -> int:
        s = 0
        if str(c.get('year')) == DEFAULT_YEAR:
            s += 100
        if c.get('site') == 'All':
            s += 50
        region = c.get('region')
        if not region or (isinstance(region, list) and len(region) == 0):
            s += 25
        return s

    combos.sort(key=score, reverse=True)
    return combos


def _run_pool(
    items: list[Any],
    concurrency: int,
    worker: Callable[[Any, int], None],
) -> None:
    index = 0
    index_lock = threading.Lock()

    def next_worker() -> None:
        nonlocal index
        while True:
            with index_lock:
                if index >= len(items):
                    return
                i = index
                index += 1
            worker(items[i], i)

    with ThreadPoolExecutor(max_workers=min(concurrency, len(items) or 1)) as pool:
        futures = [pool.submit(next_worker) for _ in range(min(concurrency, len(items) or 1))]
        for fut in as_completed(futures):
            fut.result()


def _default_filter_combo() -> dict[str, Any]:
    return {
        'year': DEFAULT_YEAR,
        'site': 'All',
        'region': [],
        'timeframe': 'FY',
    }


def warm_default_combo() -> None:
    """Priority warmup for FY default — keeps first page visit fast and avoids combo fan-out."""
    if not preload_enabled() or not sql_configured():
        return
    combo = _default_filter_combo()
    if get_fresh_metrics_bundle(combo):
        print('[preload] default FY combo already fresh in memory', flush=True)
        return
    try:
        print(f'[preload] warming default FY {DEFAULT_YEAR} combo…', flush=True)
        refresh_metrics_bundle(combo)
        print(f'[preload] default FY {DEFAULT_YEAR} combo ready', flush=True)
    except Exception as exc:
        print(f'[preload] default combo warmup failed: {exc}', flush=True)


def run_preload_cycle() -> None:
    global _status

    if not preload_enabled() or not sql_configured():
        return

    with _status_lock:
        if _status['running']:
            print('[preload] cycle already running — skip', flush=True)
            return
        _status['running'] = True
        _status['lastRunStartedAt'] = time.time() * 1000
        _status['lastRunError'] = None
        _status['combinationsDone'] = 0
        _status['combinationsSkipped'] = 0

    try:
        combos = _load_filter_combinations()
        with _status_lock:
            _status['combinationsTotal'] = len(combos)
        print(
            f'[preload] starting cycle — {len(combos)} combinations, concurrency={CONCURRENCY}',
            flush=True,
        )

        def worker(combo: dict[str, Any], _idx: int) -> None:
            if get_fresh_metrics_bundle(combo):
                with _status_lock:
                    _status['combinationsSkipped'] += 1
                    _status['combinationsDone'] += 1
                return
            try:
                def on_progress(msg: str) -> None:
                    with _status_lock:
                        done = _status['combinationsDone']
                    if done % 10 == 0:
                        total = _status['combinationsTotal']
                        print(f'[preload] {done}/{total} — {msg}', flush=True)

                refresh_metrics_bundle(combo, on_progress)
            except Exception as e:
                print(
                    f'[preload] combo failed: {json.dumps(combo)[:80]} {e}',
                    flush=True,
                )
            with _status_lock:
                _status['combinationsDone'] += 1

        _run_pool(combos, CONCURRENCY, worker)

        with _status_lock:
            _status['memoryCacheEntries'] = get_memory_cache_stats()['entries']
            _status['lastRunFinishedAt'] = time.time() * 1000
            started = _status['lastRunStartedAt'] or 0
            finished = _status['lastRunFinishedAt'] or 0
            elapsed = (finished - started) / 1000
            print(
                f'[preload] cycle complete — {_status["combinationsDone"]}/{_status["combinationsTotal"]} '
                f'({_status["combinationsSkipped"]} skipped fresh) in {elapsed:.0f}s, '
                f'{_status["memoryCacheEntries"]} memory keys',
                flush=True,
            )
    except Exception as e:
        with _status_lock:
            _status['lastRunError'] = str(e)
            _status['lastRunFinishedAt'] = time.time() * 1000
        print(f'[preload] cycle failed: {_status["lastRunError"]}', flush=True)
    finally:
        with _status_lock:
            _status['running'] = False
            _status['memoryCacheEntries'] = get_memory_cache_stats()['entries']
        _schedule_next_run()


def _schedule_next_run() -> None:
    global _next_run_timer

    if not _scheduler_started or not preload_enabled() or not sql_configured():
        return

    if _next_run_timer:
        _next_run_timer.cancel()

    with _status_lock:
        _status['nextRunAt'] = time.time() * 1000 + INTERVAL_MS

    def tick() -> None:
        run_preload_cycle()

    _next_run_timer = threading.Timer(INTERVAL_MS / 1000, tick)
    _next_run_timer.daemon = True
    _next_run_timer.start()


def stop_preload_scheduler() -> None:
    """Cancel scheduled preload cycles and prevent new ones until restart."""
    global _scheduler_started, _next_run_timer

    if _next_run_timer:
        _next_run_timer.cancel()
        _next_run_timer = None

    _scheduler_started = False
    with _status_lock:
        _status['enabled'] = False
        _status['nextRunAt'] = None
    print('[preload] scheduler stopped', flush=True)


def start_preload_scheduler() -> None:
    global _scheduler_started

    if _scheduler_started or not preload_enabled() or not sql_configured():
        return

    _scheduler_started = True
    with _status_lock:
        _status['enabled'] = True
    warm_memory_cache_from_disk()
    print(
        f'[preload] scheduler started — default warmup now, full cycle every '
        f'{INTERVAL_MS / 60000} min (after {FULL_CYCLE_DELAY_MS / 60000} min delay), '
        f'concurrency={CONCURRENCY}',
        flush=True,
    )

    def _boot() -> None:
        warm_default_combo()
        if FULL_CYCLE_DELAY_MS > 0:
            print(
                f'[preload] deferring full combo cycle for {FULL_CYCLE_DELAY_MS / 60000:.0f} min',
                flush=True,
            )
            time.sleep(FULL_CYCLE_DELAY_MS / 1000)
        run_preload_cycle()

    threading.Thread(target=_boot, daemon=True, name='preload-boot').start()


def get_preload_status() -> PreloadStatus:
    with _status_lock:
        return {
            **_status,
            'enabled': preload_enabled(),
            'interval_minutes': INTERVAL_MS / 60000,
            'concurrency': CONCURRENCY,
            'memoryCacheEntries': get_memory_cache_stats()['entries'],
        }
